### ISSUE TITLE
Undetermined shared module version when importing a file from the module directly

### ISSUE THREAD

**Issue by Hornwitser** [webpack/webpack#13457]
# Bug report

**What is the current behavior?**

When importing a file directly from a module (i.e. `import "react-dom/test-utils"`) when it's configured to be shared via ModuleFedrationPlugin a warning with a nonsensical instruction may be emitted when running a build:

```
WARNING in shared module react-dom
No required version specified and unable to automatically determine one. Unable to find required version for "react-dom" in description file (/home/user/test/node_modules/react-dom/package.json). It need to be in dependencies, devDependencies or peerDependencies.
```

**If the current behavior is a bug, please provide the steps to reproduce.**

Using the following webpack config

```js
"use strict";
const path = require("path");
const webpack = require("webpack");


module.exports = (env = {}) => ({
    entry: "./index.js",
    output: {
        publicPath: "auto",
        filename: "bundle.js",
        path: path.resolve(__dirname, "dist"),
    },
    mode: "production",
    plugins: [
        new webpack.container.ModuleFederationPlugin({
            name: "test",
            shared: {
                "react-dom": { singleton: true },
            },
        }),
    ],
});
```

And an index.js file with the content
```js
import "react-dom/test-utils"
```

Along with a package.json file with the dependencies described in reproduces the behaviour. See the complete example in [this repository](https://github.com/Hornwitser/webpack-bug).

**What is the expected behavior?**

I expect either the version of react-dom to be correctly deduced from package.json or a sensible warning/error message.

**Other relevant information:**
webpack version:  5.37.1
Node.js version: 16.0.0
Operating System: Arch Linux
Additional tools:

**Comment by alexander-akait** [comment]
Expected, you want share `react-dom`, but request `react-dom/test-utils`

**Comment by Hornwitser** [comment]
Maybe, but in my case it was a dependency of a dependency that imported `react-dom/test-utils` and I can't deduce that from the error message.

**Comment by arrgson** [comment]
I have the same issue happening with @angular/common .

`shared module @angular/common - Warning: No required version specified and unable to automatically determine one. Unable to find required version for "@angular/common" in description file (.../node_modules/@angular/common/package.json). It need to be in dependencies, devDependencies or peerDependencies.`

In the package.json file there is no entry for @angular/common under those properties.

**Comment by brendonco** [comment]
@sokra does webpack version 5.41.1 fix this issue?

**Comment by alexander-akait** [comment]
No, you can send a fix

**Comment by flyyuan** [comment]
same problem.

**Comment by deepak010negi** [comment]
Getting similar issue, any fixes?

WARNING in shared module react
No required version specified and unable to automatically determine one. Unable to find required version for "react" in description file (/Users/deepaknegi/Desktop/projects/email-template-ui/node_modules/@emotion/react/jsx-runtime/package.json). It need to be in dependencies, devDependencies or peerDependencies.

**Comment by kirill-konshin** [comment]
Possibly related https://github.com/mui-org/material-ui/issues/24013

**Comment by vugar005** [comment]
@arrgson It is because of secondary point.
 According to manfred amazing book.
The reason for this is the secondary entry point ```@angular/common/http``` which is a bit like an npm
package within an npm package. Technically, it’s just another file exposed by the npm package
```@angular/common``` .
Unsurprisingly, ```@angular/common/http``` uses ```@angular/common``` and webpack recognizes this. For this
reason, webpack wants to find out which version of ```@angular/common``` is used. For this, it looks into
the npm package’s package.json ( @angular/common/package.json ) and browses the dependencies
there. However, ```@angular/common``` itself is not a depdency of ```@angular/common``` and hence,the version
cannot be found.
You will have the same challenge with other packages using secondary entry points,e.g. @angular/material .
To avoid this situation, you can assign versions to all shared libraries **manually**:
```
 shared: {
  "@angular/core": {
   singleton: true,
   strictVersion: true,
   requiredVersion: '12.0.0'
 },
   "@angular/common": {
   singleton: true,
   strictVersion: true,
   requiredVersion: '12.0.0'
 },
   "@angular/common/http": {
   singleton: true,
   strictVersion: true,
   requiredVersion: '12.0.0'
 },
   "@angular/router": {
   singleton: true,
   strictVersion: true,
   requiredVersion: '12.0.0'
 },

 },
```

**Comment by tgelu** [comment]
> To avoid this situation, you can assign versions to all shared libraries manually

The issue with this seems to be that you lose the valuable warnings thrown from non-matching shared singletons.
For example, package X which you depend on requires v1 of dependency A but your build says requiredVersion for dependecy A is v2, then in this case you would not get that warning that versions mismatch because the required version is not determined anymore by package X but by your static "v2" in the module federation's config.
Seems to me like this is not a satisfactory solution. It wasn't in my case at least.

**Comment by EricWittmann** [comment]
There seems to be another variation of this issue for certain dependencies, such as `yaml`.  See this PR for further discussion:

https://github.com/eemeli/yaml/pull/396

And a reproducer project here:

https://github.com/EricWittmann/module-federation-with-yaml-issue

Would be interested to know if this is something webpack should be handling - e.g. by searching harder for a parent `package.json` file under certain circumstances?

**Comment by vankop** [comment]
> e.g. by searching harder for a parent package.json file under certain circumstances?

from https://docs.npmjs.com/creating-a-package-json-file 
```
package.json fields:
Required name and version fields
```
So basically this is always required fields.

**Comment by Hornwitser** [comment]
I fail to see how this issue has been resolved. Using the reproduction repository linked in the first post with webpack v5.73.0 and webpack-cli v4.9.2 I still get the following nonsensical warning message

```
WARNING in shared module react-dom
No required version specified and unable to automatically determine one. Unable to find required version for "react-dom" in description file (~/webpack-bug/node_modules/react-dom/package.json). It need to be in dependencies, devDependencies or peerDependencies.
```

It's telling me it's unable to find the version of `react-dom` _in `react-dom`'s package.json file_. It does not make any sense for `react-dom` to depend on itself, the warning is therefore nonsensical regardless of whether the behaviour of not being able to find the required version is expected or not.

**Comment by vankop** [comment]
ok, webpack should implement for shared modules package.json more smart resolving..

**Comment by eemeli** [comment]
@vankop Should this issue be reopened, or are you filing a separate issue for the planned change?

**Comment by miteokaku** [comment]
Спасибо 🙏

**Comment by ach206** [comment]
following

**Comment by kenatwex** [comment]
well Im getting byproducts of this same issue from other stuff not being articulated properly under the hood.

any workaround?

```
main.ts:22 Error: Unsatisfied version 0 from my-host of shared singleton module @apollo/client/core (required ^3.7.1)
    at getStrictSingletonVersion (consumes:49:1)
    at consumes:130:1
```

**Comment by kenatwex** [comment]
attempted this one...

```
const coreLibraries = new Set([
  '@angular/animations: { singleton: true, strictVersion: true, eager: true, requiredVersion: "^14.2.7" }',
  '@angular/cdk: { singleton: true, strictVersion: true, eager: true, requiredVersion: "^14.2.7" }',
  '@angular/common: { singleton: true, strictVersion: true, eager: true, requiredVersion: "^14.2.7" }',
  '@angular/compiler: { singleton: true, strictVersion: true, eager: true, requiredVersion: "^14.2.7" }',
  '@angular/core: { singleton: true, strictVersion: true, eager: true, requiredVersion: "^14.2.7" }',
  '@angular/forms: { singleton: true, strictVersion: true, eager: true, requiredVersion: "^14.2.7" }',
  '@angular/router: { singleton: true, strictVersion: true, eager: true, requiredVersion: "^14.2.7" }',
  '@apollo/client: { singleton: true, strictVersion: true, eager: true, requiredVersion: "^3.7.1" }',

  '@ngneat/transloco: { singleton: true, strictVersion: true, eager: true, requiredVersion: "^4.1.1" }',
  '@ngrx/component: { singleton: true, strictVersion: true, eager: true, requiredVersion: "^14.3.2" }',
  '@ngrx/component-store: { singleton: true, strictVersion: true, eager: true, requiredVersion: "^14.3.2" }',
  '@ngrx/effects: { singleton: true, strictVersion: true, eager: true, requiredVersion: "^14.3.2" }',
  '@ngrx/entity: { singleton: true, strictVersion: true, eager: true, requiredVersion: "^14.3.2" }',
  '@ngrx/router-store: { singleton: true, strictVersion: true, eager: true, requiredVersion: "^14.3.2" }',
  '@ngrx/store: { singleton: true, strictVersion: true, eager: true, requiredVersion: "^14.3.2" }',
  '@ngrx/store-devtools: { singleton: true, strictVersion: true, eager: true, requiredVersion: "^14.3.2" }',

  'apollo-angular: { singleton: true, strictVersion: true, eager: true, requiredVersion: "^4.1.0" }',
  'graphql: { singleton: true, strictVersion: true, eager: true, requiredVersion: "^16.6.0" }',
]);

module.exports = {
  name: 'host-yeti',
  remotes: ['ui-auth'],
  additionalShared: [
    '@angular/animations'
  ],
  shared: (libraryName, defaultConfig) => {
    if (coreLibraries.has(libraryName)) {
      return defaultConfig;
    }

    // Returning false means the library is not shared.
    return false;
  },
};
```

terminal warning
`shared module @apollo/client/core -> /Users/me/PR/150305/my/node_modules/@apollo/client/core/index.js - Warning: No version specified and unable to automatically determine one. No version in description file (usually package.json). Add version to description file /Users/me/PR/150305/my/node_modules/@apollo/client/core/package.json, or manually specify version in shared config.`

back to same thing

browser console
```
Error: Unsatisfied version 0 from host-yeti of shared singleton module @apollo/client/core (required ^3.7.1)
    at getStrictSingletonVersion (consumes:49:1)
```

**Comment by muhammedMoussa** [comment]
following

**Comment by alexander-akait** [comment]
bump

**Comment by akropolis7** [comment]
You must add version into shared package!
before: `shared: { 'react': { singleton: true }, 'react-dom': { singleton: true } }`
after: `shared: { 'react@^18.2.0': { singleton: true }, 'react-dom@^18.2.0': { singleton: true } }`

**Comment by tgelu** [comment]
> You must add version into shared package! before: `shared: { 'react': { singleton: true }, 'react-dom': { singleton: true } }` after: `shared: { 'react@^18.2.0': { singleton: true }, 'react-dom@^18.2.0': { singleton: true } }`

I think that has been [discussed above](https://github.com/webpack/webpack/issues/13457#issuecomment-1110952016).

**Comment by lordrip** [comment]
Hi @eemeli, I've been reading your comments in regards `yaml` and I understand that this is something that we would need to tackle from `webpack` side but I just wanted to ask, from your point of view, is there a workaround that we could do in the consumer's modules in the meantime?

If that's not the case, no worries, thanks in advance :v:

**Comment by lordrip** [comment]
Manually adding name and version property to `yaml/browser/package.json` seems to work, but according to @eemeli this is not the way to go

```json
{
  "name": "yaml", //<--- added
  "version": "2.2.2", //<--- added
  "type": "module"
}
```

**Comment by eemeli** [comment]
@lordrip I'm not aware of a workaround; this bug still needs to get fixed in `webpack`. I've suggested one possible solution above.

I'm not myself using Webpack at the moment, so would recommend that someone who's directly affected submit a PR fixing this.

**Comment by krutoo** [comment]
> Manually adding name and version property to `yaml/browser/package.json` seems to work, but according to @eemeli this is not the way to go
> 
> ```json
> {
>   "name": "yaml", //<--- added
>   "version": "2.2.2", //<--- added
>   "type": "module"
> }
> ```

I tried this but it is not wirking, warnings about shared dependencies still in webpack serve log

**Comment by MegaDhakse2** [comment]
i am also getting this warning on upgrading @sentry/react version from v7.x to v8.x.. Application working fine but this warning is disturbing.. Why this occuring first of all.. i have installed from a source and migrated following their documentation, despite it showing failed to see a version.. Gosh!!!!
![sentry_version_upgrading](https://github.com/user-attachments/assets/ce4db48b-0e30-4190-bda9-5d7a2bbeb240)

**Comment by alexander-akait** [comment]
@MegaDhakse2 Please provide a reproducible repository using github, thank you

**Comment by krutoo** [comment]
Can we reopen this issue or open new? The problem with dual ESM+CJS packages is still here

**Comment by alexander-akait** [comment]
@krutoo Can you create a reproducible test repo, need investigate

**Comment by Bairon29** [comment]
I had the same issue with “deep-object-diff” with this. Noticed the required version and version.
`new ModuleFederationPlugin({
…
shared: {
…packageJSONDeps,
“deep-object-diff”: {
singleton: true,
version: packageJSONDeps[“deep-object-diff”],
requiredVersion: packageJSONDeps[“deep-object-diff”]
}}}`

**Comment by MegaDhakse2** [comment]
Hi all, 
I found a lead.. Someone asks for a demo repo or reproductive steps. For that i have integrated in a standalone project to reproduce that error. But it woks fine without warning there.

Then only i found that, this is only occurs in **Module Fedaration apps**. For standalone alone apps it works fine..

So can anyone HELP me to get out from this..

**Comment by krutoo** [comment]
> @krutoo Can you create a reproducible test repo, need investigate

@alexander-akait here is the example of problem:
https://github.com/krutoo/module-federation-dual-packages

**Comment by alexander-akait** [comment]
@krutoo Yeah, because there is a bug in `@sentry/browser`, look at `node_modules/@sentry/browser/build/npm/esm/package.json`:

```json
{"type":"module","sideEffects":false}
```

 No `name` and `version`.
 
 From docs:
 
 > If you plan to publish your package, the most important things in your package.json are the name and version fields as they will be required.
 
 > If you plan to publish your package, the most important things in your package.json are the name and version fields as they will be required.
 
 https://docs.npmjs.com/cli/v10/configuring-npm/package-json
 
 The package is published, and it is necessary to specify these fields, even if it is used for switching the `type`.
 
 Thear already have:
 ```
 "exports": {
    "./package.json": "./package.json",
    ".": {
      "import": {
        "types": "./build/npm/types/index.d.ts",
        "default": "./build/npm/esm/index.js"
      },
      "require": {
        "types": "./build/npm/types/index.d.ts",
        "default": "./build/npm/cjs/index.js"
      }
    }
  },
 ```
 
 Therefore, the presence of an additional files is not even required, but can only break the logic of determining versions and names. 
 
 I don't think we are making a mistake here, rather the opposite, we are allowing this problem to be caught. The only solution is to specify the version manually, you can read the `package.json` of `@sentry/browser` so you don't have to write it manually. 
 
 I would advise you to open an issue in this repository and refer to this issue, such problems will most likely appear in any bundlers (which support module federation), and even most likely in other tools.

**Comment by krutoo** [comment]
@alexander-akait but name and version is defined in root package.json

according to this logic, all information from the root package.json should be duplicated

I did this in my packages and specified the version, but then the federation module complained about the lack of dependency versions

**Comment by alexander-akait** [comment]
@krutoo 

> according to this logic, all information from the root package.json should be duplicated

No, `package.json` do not inherit information from the previous one

Any `package.json` should have `name` and `version` if it is published, even for nested, this is the rule

**Comment by krutoo** [comment]
@alexander-akait `@sentry/node` has exactly same structure and same nested package.json, when you import it in Node.js .mjs file it works fine

I think the rule you are talking about applies only to the main `package.json`

**Comment by eemeli** [comment]
The npm docs on this are misleading, and unfortunately my attempt to clarify them was recently closed by @wraithgar: npm/documentation#134.

**Comment by alexander-akait** [comment]
@eemeli - Yeah I see, because changing the resolver logic can be dangerous without spec/docs and because developers close this issue I would interpret this as mandatory fields always for any published projects
